go/services/adr/client.go

```diff
@@ -0,0 +1,539 @@
+	c.logger.Debug(
+		"Observing asset updates",
+		"aepName",
+		aepName,
+		"assetName",
+		assetName,
+	)
+
+	req := adrbaseservice.NotifyOnAssetUpdateRequestPayload{
+		NotificationRequest: adrbaseservice.NotifyOnAssetUpdateRequestSchema{
+			AssetName:               assetName,
+			NotificationMessageType: adrbaseservice.On,
+		},
+	}
+
+	tokens := map[string]string{
+		aepNameTokenKey: aepName,
+	}
+
+	// Use the appropriate command invoker
+	resp, err := c.notifyOnAssetUpdateInvoker.NotifyOnAssetUpdate(
+		ctx,
+		req,
+		protocol.WithTopicTokens(tokens),
+	)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	key := aepName + "~" + assetName
+	c.mu.Lock()
+	c.observedAssets[key] = struct{}{}
+	c.mu.Unlock()
+
+	return &resp.Payload.NotificationResponse, nil
+}
+
+// UnobserveAssetUpdates stops observation of asset updates.
+func (c *Client) UnobserveAssetUpdates(
+	ctx context.Context,
+	aepName, assetName string,
+) (*NotificationResponse, error) {
+	c.logger.Debug(
+		"Unobserving asset updates",
+		"aepName",
+		aepName,
+		"assetName",
+		assetName,
+	)
+
+	req := adrbaseservice.NotifyOnAssetUpdateRequestPayload{
+		NotificationRequest: adrbaseservice.NotifyOnAssetUpdateRequestSchema{
+			AssetName:               assetName,
+			NotificationMessageType: adrbaseservice.Off,
+		},
+	}
+
+	tokens := map[string]string{
+		aepNameTokenKey: aepName,
+	}
+
+	resp, err := c.notifyOnAssetUpdateInvoker.NotifyOnAssetUpdate(
+		ctx,
+		req,
+		protocol.WithTopicTokens(tokens),
+	)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	key := aepName + "~" + assetName
+	c.mu.Lock()
+	delete(c.observedAssets, key)
+	c.mu.Unlock()
+
+	return &resp.Payload.NotificationResponse, nil
+}
+
+// GetAssetEndpointProfile retrieves an asset endpoint profile by name.
+func (c *Client) GetAssetEndpointProfile(
+	ctx context.Context,
+	aepName string,
+) (*AssetEndpointProfile, error) {
+	c.logger.Debug("Getting asset endpoint profile", "aepName", aepName)
+
+	tokens := map[string]string{
+		aepNameTokenKey: aepName,
+	}
+
+	resp, err := c.getAssetEndpointProfileCommandInvoker.GetAssetEndpointProfile(
+		ctx,
+		protocol.WithTopicTokens(tokens),
+	)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	return &resp.Payload.AssetEndpointProfile, nil
+}
+
+// UpdateAssetEndpointProfileStatus updates the status of an asset endpoint profile.
+func (c *Client) UpdateAssetEndpointProfileStatus(
+	ctx context.Context,
+	aepName string,
+	status *AssetEndpointProfileStatus,
+) (*AssetEndpointProfile, error) {
+	c.logger.Debug("Updating asset endpoint profile status", "aepName", aepName)
+
+	req := adrbaseservice.UpdateAssetEndpointProfileStatusRequestPayload{
+		AssetEndpointProfileStatusUpdate: *status,
+	}
+
+	tokens := map[string]string{
+		aepNameTokenKey: aepName,
+	}
+
+	resp, err := c.updateAssetEndpointProfileStatusCommandInvoker.UpdateAssetEndpointProfileStatus(
+		ctx,
+		req,
+		protocol.WithTopicTokens(tokens),
```

Comment:
nit: If you're not reusing the `tokens` variable anywhere else, you can actually do `protocol.WithTopicTokens{ aepNameTokenKey: aepName }`. That was one of the benefits of using a type over the traditional functional option.

Reply:
Properly implemented token

Reply:
Got them all this time